### PR TITLE
bpo-32911: Remove the docstring attribute of AST types

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is based on #5927 and goes one step further to Python 3.6. It doesn't introduce a new AST type DocString.

https://bugs.python.org/issue32911

### CODE REVIEW

**Review by ncoghlan — CHANGES_REQUESTED** [review]
Thanks for this Serhiy. In addition to the b5 NEWS entry, the only other item I see missing here is a magic number bump for the bytecode (so any debug and opt-level 1 bytecode generated with the earlier alphas and betas gets regenerated with docstrings included again).

**Comment by ncoghlan** [comment]
Note that while the previous 3.7.0a1 NEWS entry is modified by the current patch, there'll need to be a dedicated NEWS entry for the reversion that will then appear under 3.7.0b5.

File: Python/ast_opt.c [hunk 1]
```diff
@@ -452,12 +452,51 @@ static int astfold_excepthandler(excepthandler_ty node_, PyArena *ctx_, int opti
...
+}
+
+static int
+astfold_body(asdl_seq *stmts, PyArena *ctx_, int optimize_)
+{
+    if (!asdl_seq_LEN(stmts)) {
+        return 1;
+    }
+    int docstring = isdocstring((stmt_ty)asdl_seq_GET(stmts, 0));
+    CALL_SEQ(astfold_stmt, stmt_ty, stmts);
```
Comment (methane) [inline Python/ast_opt.c:474]: Loop from 1 seems simpler than JoinedStr hack.

```c
    for (int i = docstring; i < asdl_seq_LEN(stmts); i++) {
        stmt_ty st = (stmt_ty)asdl_seq_GET(stmts, i);
        if (st != NULL && !astfold_stmt(st, ctx_, optimize_)) {
            return 0;
        }
    }
```

**Review by methane — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Sorry, I don't know Git well still, and loss a news entry when applied Inada's patch.

I didn't think that a magic number bumping is needed, because this PR affects AST. But it also affects the first line number of nodes with docstrings. I'll bump a magic number.

File: Python/ast_opt.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Python/ast_opt.c:474]: A JoinedStr hack is needed for preventing interpreting a new constant string created by optimization as a docstring:
```python
def foo():
    'Not a' + ' docstring'
    pass
```
but we still want to apply optimizations to the first statement.

File: Misc/NEWS.d/next/Core and Builtins/2018-02-27-20-57-00.bpo-32911.cmKfco.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Reverted :issue:`29463`. ``docstring`` field is removed from Module, ClassDef,
+FunctionDef, and AsyncFunctionDef ast nodes which is added in 3.7a1.  Docstring
```
Comment (ned-deily) [inline Misc/NEWS.d/next/Core and Builtins/2018-02-27-20-57-00.bpo-32911.cmKfco.rst]: "which was added"

File: Misc/NEWS.d/next/Core and Builtins/2018-02-27-20-57-00.bpo-32911.cmKfco.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Reverted :issue:`29463`. ``docstring`` field is removed from Module, ClassDef,
```
Comment (ned-deily) [inline Misc/NEWS.d/next/Core and Builtins/2018-02-27-20-57-00.bpo-32911.cmKfco.rst]: We should say why we are doing this.  Perhaps start the entry with:
"Due to unexpected compatibility issues discovered during downstream beta testing, reverted ... "

**Review by ned-deily — CHANGES_REQUESTED** [review]
A couple of wording improvements

**Comment by serhiy-storchaka** [comment]
I have made the requested changes; please review again.

**Review by ned-deily — APPROVED** [review]

**Comment by ncoghlan** [comment]
Thanks @serhiy-storchaka!

**Pull request merged**